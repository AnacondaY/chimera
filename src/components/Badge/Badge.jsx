import React from 'react';
import BaseComponent from '../../base/BaseComponent';
import PropTypes from '../../base/prop-types';

export default class Badge extends BaseComponent {

    static propTypes = {
        number: PropTypes.number,
        dot: PropTypes.bool,
        blink: PropTypes.bool,
        maximum: PropTypes.number,
        showZero: PropTypes.bool,
        maxTemplate: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
        type: PropTypes.oneOf(['primary', 'default', 'error', 'warning', 'success', 'info'])
    }

    static defaultProps = {
        dot: false,
        blink: false,
        maximum: 99,
        showZero: false,
        type: 'error',
    }

    renderNumber(): React.ReactNode{
        const { number, maximum, maxTemplate, showZero } = this.props;
        if(number <= maximum){
            return number;
        }else{
            if(maxTemplate){
                if(typeof maxTemplate === 'function'){
                    return maxTemplate(number, maximum);
                }else{
                    return maxTemplate;
                }
            }
            return `${maximum}+`;
        }
    }

    render(){
        const { number, maximum, showZero, maxTemplate, dot, type, blink, children } = this.props;
        const symbolNode = (
            <span className={this.classNames({
                'cmr-badge-symbol': true,
                'cmr-badge-symbol-dot': dot,
                'cmr-badge-symbol-blink': dot && blink
            })}>
                {!dot && this.renderNumber()}
            </span>
        );
        return (
            <span className={this.classNames({
                'cmr-badge': true,
                [`cmr-badge-${type}`]: true
            })}>
                {children}
                { number === 0 ? ( showZero ? symbolNode : null) : symbolNode  }
            </span>
        );
    }

}