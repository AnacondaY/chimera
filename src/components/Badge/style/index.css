@keyframes blink {
  0% {
    transform: scale(0.75);
    opacity: .6; }
  100% {
    transform: scale(2.25);
    opacity: 0; } }

.cmr-badge {
  display: inline-block;
  position: relative; }
  .cmr-badge-symbol {
    position: absolute;
    top: 0;
    right: 0;
    height: 18px;
    line-height: 18px;
    padding: 0 4px;
    text-align: center;
    min-width: 18px;
    border-radius: 10px;
    color: #fff;
    transform: translate(50%, -50%);
    font-size: 12px;
    white-space: nowrap;
    box-sizing: border-box; }
    .cmr-badge-symbol-dot {
      border-radius: 50%;
      height: 10px;
      width: 10px;
      min-width: 10px;
      padding: 0; }
  .cmr-badge-primary .cmr-badge-symbol {
    background: #1b98f3; }
    .cmr-badge-primary .cmr-badge-symbol-blink:after {
      content: '';
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      border-radius: 50%;
      border: 1px solid #1b98f3;
      animation: blink 1.25s infinite backwards ease-out; }
  .cmr-badge-success .cmr-badge-symbol {
    background: #20b96d; }
    .cmr-badge-success .cmr-badge-symbol-blink:after {
      content: '';
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      border-radius: 50%;
      border: 1px solid #20b96d;
      animation: blink 1.25s infinite backwards ease-out; }
  .cmr-badge-info .cmr-badge-symbol {
    background: #22ccc2; }
    .cmr-badge-info .cmr-badge-symbol-blink:after {
      content: '';
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      border-radius: 50%;
      border: 1px solid #22ccc2;
      animation: blink 1.25s infinite backwards ease-out; }
  .cmr-badge-error .cmr-badge-symbol {
    background: #f05a5a; }
    .cmr-badge-error .cmr-badge-symbol-blink:after {
      content: '';
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      border-radius: 50%;
      border: 1px solid #f05a5a;
      animation: blink 1.25s infinite backwards ease-out; }
  .cmr-badge-default .cmr-badge-symbol {
    background: #000; }
    .cmr-badge-default .cmr-badge-symbol-blink:after {
      content: '';
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      border-radius: 50%;
      border: 1px solid #000;
      animation: blink 1.25s infinite backwards ease-out; }
  .cmr-badge-warning .cmr-badge-symbol {
    background: #fac525; }
    .cmr-badge-warning .cmr-badge-symbol-blink:after {
      content: '';
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      border-radius: 50%;
      border: 1px solid #fac525;
      animation: blink 1.25s infinite backwards ease-out; }
