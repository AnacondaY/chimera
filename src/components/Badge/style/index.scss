@import '../../style/main/base.scss';
@import '../../style/theme/default.scss';
@import './mixin.scss';

$badge: #{$class-prefix}-badge;

.#{$badge}{
    display: inline-block;
    position: relative;
    &-symbol{
        position: absolute;
        top: 0;
        right: 0;
        height: 18px;
        line-height: 18px;
        padding: 0 4px;
        text-align: center;
        min-width: 18px;
        border-radius: 10px;
        color: $default;
        transform: translate(50%, -50%);
        font-size: 12px;
        white-space: nowrap;
        box-sizing: border-box;
        &-dot{
            border-radius: 50%;
            height: 10px;
            width: 10px;
            min-width: 10px;
            padding: 0;
        }
    }
    &-primary{
        @include symbol-generator($primary, $badge);
    }
    &-success{
        @include symbol-generator($success, $badge);
    }
    &-info{
        @include symbol-generator($info, $badge);
    }
    &-error{
        @include symbol-generator($error, $badge);
    }
    &-default{
        @include symbol-generator($inverse, $badge);
    }
    &-warning{
        @include symbol-generator($warning, $badge);
    }
}