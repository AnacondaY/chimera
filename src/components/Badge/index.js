import Badge from './Badge';
//import './style';

export default Badge;