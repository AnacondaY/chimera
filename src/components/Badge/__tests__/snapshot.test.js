import Badge from '../';

describe('===== Badge Snapshot Test =====', () => {
    
    test('renders corrently', () => {
        const badge = create(
            <Badge number={24} />
        ).toJSON();
        expect(badge).toMatchSnapshot();
    });

    test('renders corrently with number greater than maximun', () => {
        const badge = create(
            <Badge number={100} />
        ).toJSON();
        expect(badge).toMatchSnapshot();
    });

});