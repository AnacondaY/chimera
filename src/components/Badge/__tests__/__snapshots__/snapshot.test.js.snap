// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`===== Badge Snapshot Test ===== renders corrently 1`] = `
<span
  className="cmr-badge cmr-badge-error"
>
  <span
    className="cmr-badge-symbol"
  >
    24
  </span>
</span>
`;

exports[`===== Badge Snapshot Test ===== renders corrently with number greater than maximun 1`] = `
<span
  className="cmr-badge cmr-badge-error"
>
  <span
    className="cmr-badge-symbol"
  >
    99+
  </span>
</span>
`;
